File: frontend/src/types/security.ts
export type SecuritySeverityLevel = 'CRITICAL' | 'HIGH' | 'MEDIUM' | 'LOW'
export type SecurityConfidenceLevel = 'HIGH' | 'MEDIUM' | 'LOW'

export type SecurityIssueType =
  | 'hardcoded_secret'
  | 'sql_injection'
  | 'command_injection'
  | 'dangerous_eval'
  | 'insecure_config'
  | 'weak_crypto'
  | 'insecure_password'

export type SecurityCategory =
  | 'SECRETS'
  | 'INJECTION'
  | 'DANGEROUS_CALLS'
  | 'CONFIGURATION'
  | 'CRYPTO'
  | 'AUTHENTICATION'
  | string

export interface SecurityIssue {
  id: string
  analysis_id: string
  project_id: string
  category: string
  issue_type: SecurityIssueType | string
  severity: SecuritySeverityLevel
  confidence: SecurityConfidenceLevel
  file_path: string
  line_number: number
  end_line: number | null
  symbol_name: string | null
  message: string
  description: string
  evidence: string | null
  recommendation: string | null
  created_at: string
}

export interface SecurityMetrics {
  files_audited: number
  files_skipped: number
  by_category: Record<string, number>
  by_type: Record<string, number>
  by_severity: Record<string, number>
}

export interface SecurityAnalysis {
  id: string
  project_id: string
  scan_id: string
  status: 'PENDING' | 'IN_PROGRESS' | 'COMPLETED' | 'FAILED'
  started_at: string
  completed_at: string | null
  total_issues: number
  critical_count: number
  high_count: number
  medium_count: number
  low_count: number
  metrics: SecurityMetrics
  error_message: string | null
}

export interface SecuritySnippetLine {
  line_number: number
  content: string
  is_highlighted: boolean
}

export interface SecurityCodeSnippet {
  file_path: string
  start_line: number
  end_line: number
  target_line: number
  lines: SecuritySnippetLine[]
}

File: backend/app/models/security.py
from sqlalchemy import Column, String, Integer, Text, ForeignKey, DateTime, JSON
from sqlalchemy.orm import relationship
from backend.app.core.database import Base
from backend.app.models.base import generate_uuid, utc_now


class SecurityAnalysis(Base):
    __tablename__ = "security_analyses"

    id = Column(String(36), primary_key=True, default=generate_uuid, index=True)
    project_id = Column(String(36), ForeignKey("projects.id", ondelete="CASCADE"), nullable=False, index=True)
    scan_id = Column(String(36), ForeignKey("project_scans.id", ondelete="CASCADE"), nullable=False, index=True)

    status = Column(String(30), default="COMPLETED")  # PENDING, IN_PROGRESS, COMPLETED, FAILED
    started_at = Column(DateTime, default=utc_now)
    completed_at = Column(DateTime, nullable=True)

    # Issue Counts
    total_issues = Column(Integer, default=0)
    critical_count = Column(Integer, default=0)
    high_count = Column(Integer, default=0)
    medium_count = Column(Integer, default=0)
    low_count = Column(Integer, default=0)

    # Detailed Vulnerability Metrics (Stored as JSON)
    metrics = Column(JSON, default=dict)

    error_message = Column(Text, nullable=True)

    # Relationships
    project = relationship("Project", back_populates="security_analyses")
    scan = relationship("ProjectScan")
    issues = relationship("SecurityIssue", back_populates="analysis", cascade="all, delete-orphan")


class SecurityIssue(Base):
    __tablename__ = "security_issues"

    id = Column(String(36), primary_key=True, default=generate_uuid, index=True)
    analysis_id = Column(String(36), ForeignKey("security_analyses.id", ondelete="CASCADE"), nullable=False, index=True)
    project_id = Column(String(36), ForeignKey("projects.id", ondelete="CASCADE"), nullable=False, index=True)

    category = Column(String(50), default="security", index=True)
    issue_type = Column(String(60), nullable=False, index=True)
    # Types: "hardcoded_secret", "sql_injection", "command_injection", "dangerous_eval", "insecure_config", "weak_crypto", "insecure_password"

    severity = Column(String(20), nullable=False, index=True)
    # Levels: "CRITICAL", "HIGH", "MEDIUM", "LOW"

    confidence = Column(String(20), default="HIGH", index=True)
    # Levels: "HIGH", "MEDIUM", "LOW"

    file_path = Column(String(500), nullable=False, index=True)
    line_number = Column(Integer, nullable=False)
    end_line = Column(Integer, nullable=True)
    symbol_name = Column(String(150), nullable=True)

    message = Column(String(255), nullable=False)
    description = Column(Text, nullable=False)
    evidence = Column(Text, nullable=True)  # STRICTLY MASKED
    recommendation = Column(Text, nullable=True)

    created_at = Column(DateTime, default=utc_now)

    # Relationships
    analysis = relationship("SecurityAnalysis", back_populates="issues")
    project = relationship("Project")
